File: frontend/src/types/sites.ts
export type SiteType = 'nginx' | 'python' | 'wordpress' | 'nodejs'
export type ServiceManager = 'systemd' | 'pm2'
export type SiteAction = 'start' | 'stop' | 'restart'

export interface SiteStatus {
  status: string
  uptime: string | null
  pid: number | null
}

export interface SiteResponse {
  id: string
  name: string
  type: SiteType
  service_name: string
  service_manager: ServiceManager
  config_file_path: string | null
  log_paths: string[]
  description: string | null
  created_at: string
  updated_at: string
  status: SiteStatus | null
}

export interface SiteCreate {
  name: string
  type: SiteType
  service_name: string
  service_manager: ServiceManager
  config_file_path?: string
  log_paths?: string[]
  description?: string
}

export interface SiteUpdate extends Partial<SiteCreate> {}

export interface SiteActionResponse {
  success: boolean
  output: string
}

export interface ConfigReadResponse {
  content: string
  path: string
}

export interface NginxSiteCandidate {
  name: string
  config_file_path: string
  server_names: string[]
  log_paths: string[]
  already_exists: boolean
}

export interface NginxDiscoverResponse {
  nginx_available: boolean
  candidates: NginxSiteCandidate[]
}

export interface NginxImportResponse {
  imported: number
  skipped: number
  sites: SiteResponse[]
}

export interface TrafficBucket {
  minute: string
  count: number
}

export interface SiteTrafficResponse {
  site_id: string
  window_minutes: number
  total_requests: number
  total_bytes: number
  status_2xx: number
  status_3xx: number
  status_4xx: number
  status_5xx: number
  requests_per_minute: TrafficBucket[]
}

export interface SystemServiceResponse {
  service_name: string
  description: string | null
  load_state: string
  active_state: string
  sub_state: string
  uptime: string | null
  pid: number | null
  unit_file_state: string
  autostart_enabled: boolean
  starred: boolean
}

File: backend/app/schemas/sites.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

SiteType = Literal["nginx", "python", "wordpress", "nodejs"]
ServiceManager = Literal["systemd", "pm2"]
SiteAction = Literal["start", "stop", "restart"]


class SiteCreate(BaseModel):
    name: str
    type: SiteType
    service_name: str
    service_manager: ServiceManager
    config_file_path: str | None = None
    log_paths: list[str] = []
    description: str | None = None


class SiteUpdate(BaseModel):
    name: str | None = None
    type: SiteType | None = None
    service_name: str | None = None
    service_manager: ServiceManager | None = None
    config_file_path: str | None = None
    log_paths: list[str] | None = None
    description: str | None = None


class SiteStatus(BaseModel):
    status: str           # active | inactive | failed | unknown
    uptime: str | None = None
    pid: int | None = None


class SiteResponse(BaseModel):
    id: str
    name: str
    type: str
    service_name: str
    service_manager: str
    config_file_path: str | None
    log_paths: list[str]
    description: str | None
    created_at: datetime
    updated_at: datetime
    status: SiteStatus | None = None

    model_config = {"from_attributes": True}


class SiteActionRequest(BaseModel):
    action: SiteAction


class SiteActionResponse(BaseModel):
    success: bool
    output: str = ""


class ConfigReadResponse(BaseModel):
    content: str
    path: str


class ConfigWriteRequest(BaseModel):
    content: str
